mountpoint-s3-client/src/s3_crt_client.rs

```diff
@@ -150,11 +149,16 @@ impl S3CrtClient {
     /// object data.
     fn new_request_template(&self, method: &str, bucket: &str) -> Result<S3Message, ConstructionError> {
         let (uri, path_prefix) = self.endpoint.for_bucket(bucket)?;
-        let hostname = uri.host_name();
+        let hostname = uri.host_name().to_str().unwrap();
+        let port = uri.host_port();
+        let mut hostname_header = format!("{}", hostname);
+        if port > 0 {
+            hostname_header = format!("{}:{}", hostname, port);
+        }
```

Comment:
Just to avoid the double `format!` calls (double allocation):
```suggestion
        let hostname_header = if port > 0 {
            format!("{}:{}", hostname, port)
        } else {
            format!("{}", hostname)
        };
```